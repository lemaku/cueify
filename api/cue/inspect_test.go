package cue

import (
	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"encoding/json"
	"fmt"
	"github.com/gkampitakis/go-snaps/snaps"
	"testing"
)

func matchInspectSnapshot(t *testing.T, value interface{}) {
	jsonResult, err := json.MarshalIndent(value, "", "    ")
	if err != nil {
		panic("Error: couldn't serialize result to use for the snapshot")
	}
	snaps.MatchSnapshot(t, string(jsonResult))
}

func TestUniversitiesSchema(t *testing.T) {
	schema := `
#student: {
	matNr:  string & =~"^[0-9]{8}$"
	name:   string
	active: *true | bool
    if active {
        semester: int
    }
}

#universities: {
	tuwien: {
		name: "Vienna University of Technology",
		students: [...#student]
	},
	countryCode?: string
}

#export: #universities
`
	matchInspectSnapshot(t, Inspect([]string{}, "{}", schema))
	matchInspectSnapshot(t, Inspect([]string{"tuwien"}, `{ "tuwien": { "students": [] } }`, schema))
	matchInspectSnapshot(t, Inspect([]string{"tuwien", "students"}, `{
		"tuwien": {
			"name": "Vienna University of Technology",
			"students": [
				{
					"name": "Test"
				}
			]
		}
	}`, schema))
	matchInspectSnapshot(t, Inspect([]string{"tuwien", "students", "0"}, `{
		"tuwien": {
			"name": "Vienna University of Technology",
			"students": [
				{
					"name": "Test"
				}
			]
		}
	}`, schema))
}

func TestInspectComplexSchema(t *testing.T) {
	raw := `{
	a: {
		k1: []
	}
}`
	schema := `#export: {
	a: {
		a?: string
		b: number | bool
		c: { d: string } | string
		e: float | _
		f: null
		g: >2
		h: string
		i: "Hello, \(h)!"
		j: bytes,
		k1: [...string]
		k2: [...string]
		l: [...string] | [...int]
		m: { a: string } | { b: string }
	}
}`
	matchInspectSnapshot(t, Inspect([]string{"a"}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"a", "k1"}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"a", "k2"}, raw, schema))
}

func TestInspectSuperComplexSchema(t *testing.T) {
	raw := `{
	"b": [],
	"c": {},
	"d": [],
	"e": {}
}`
	schema := `#export: {
	b?: [...string] | { b: string }
	b2: [...string] | { b: string }
	c: [...string] | { b: string }
	c2: [...string] | { b: string }
	d: [...string] | [...int|bool]
	d2: [...string] | [...int|bool]
	e: { a: string } | { b: string }
	e2: { a: string } | { b: string }
	f: { b?: string } | { b: int }
}`
	matchInspectSnapshot(t, Inspect([]string{}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"b"}, raw, schema)) // TODO this is still a problem
	matchInspectSnapshot(t, Inspect([]string{"b2"}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"d"}, raw, schema))  // TODO still issue because .Eval can't be used
	matchInspectSnapshot(t, Inspect([]string{"d2"}, raw, schema)) // TODO still issue because .Eval can't be used
	matchInspectSnapshot(t, Inspect([]string{"e"}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"e2"}, raw, schema))
	matchInspectSnapshot(t, Inspect([]string{"f"}, raw, schema))
}

func TestLookupPath(t *testing.T) {
	context := cuecontext.New()

	// I would expect this to print "string"
	value := context.CompileString(`["abc"]`)
	fmt.Println(value.LookupPath(cue.MakePath(cue.AnyIndex)))
	fmt.Println(value.Eval().LookupPath(cue.MakePath(cue.AnyIndex)))

	// I would expect this to print "string" just like if "value" was simply [...string]
	value = context.CompileString(`([...string] | { a: string }) & []`)
	fmt.Println(value.Eval().LookupPath(cue.MakePath(cue.AnyIndex)))

	value = context.CompileString(`([...string] | [...int]) & []`)
	op, values := value.Eval().Expr()
	// op is OrOp
	fmt.Println(op)
	// => I would expect this to print "string"
	fmt.Println(values[0].Eval().LookupPath(cue.MakePath(cue.AnyIndex)))
	// => I would expect this to print "int"
	fmt.Println(values[1].Eval().LookupPath(cue.MakePath(cue.AnyIndex)))

	// I would expect this to print "string"
	value = context.CompileString(`([...string] | { a: string }) & ["abc"]`)
	fmt.Println(value.Eval().LookupPath(cue.MakePath(cue.AnyIndex)))

	// U would expect this to print "string"
	value = context.CompileString(`{ a: [...[...string]]} & {a: [[]]}`).LookupPath(cue.ParsePath("a[0]"))
	fmt.Println(value.Eval().LookupPath(cue.MakePath(cue.AnyIndex)))
}
