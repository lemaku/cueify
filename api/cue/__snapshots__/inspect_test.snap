
[TestUniversitiesSchema - 1]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "tuwien"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "countryCode"
            ],
            "type": [
                "string"
            ],
            "optional": true
        }
    ]
}
---

[TestUniversitiesSchema - 2]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "tuwien",
                "name"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "tuwien",
                "students"
            ],
            "type": [
                "list"
            ],
            "optional": false
        }
    ]
}
---

[TestInspectComplexSchema - 1]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "a",
                "a"
            ],
            "type": [
                "string"
            ],
            "optional": true
        },
        {
            "path": [
                "a",
                "b"
            ],
            "type": [
                "bool",
                "int",
                "float"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "c"
            ],
            "type": [
                "string",
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "e"
            ],
            "type": [
                "null",
                "bool",
                "int",
                "float",
                "string",
                "bytes",
                "list",
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "f"
            ],
            "type": [
                "null"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "g"
            ],
            "type": [
                "int",
                "float"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "h"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "i"
            ],
            "type": [
                "bottom"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "j"
            ],
            "type": [
                "bytes"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "k1"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "k2"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "l"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "a",
                "m"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        }
    ]
}
---

[TestUniversitiesSchema - 3]
{
    "type": [
        "list"
    ],
    "of": [
        "struct"
    ],
    "properties": [
        {
            "path": [
                "tuwien",
                "students",
                "0"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        }
    ]
}
---

[TestUniversitiesSchema - 4]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "tuwien",
                "students",
                "0",
                "matNr"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "tuwien",
                "students",
                "0",
                "name"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "tuwien",
                "students",
                "0",
                "semester"
            ],
            "type": [
                "int"
            ],
            "optional": false
        },
        {
            "path": [
                "tuwien",
                "students",
                "0",
                "active"
            ],
            "type": [
                "bool"
            ],
            "optional": false
        }
    ]
}
---

[TestInspectSuperComplexSchema - 1]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "b"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "b2"
            ],
            "type": [
                "list",
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "c"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "c2"
            ],
            "type": [
                "list",
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "d"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "d2"
            ],
            "type": [
                "list"
            ],
            "optional": false
        },
        {
            "path": [
                "e"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "e2"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        },
        {
            "path": [
                "f"
            ],
            "type": [
                "struct"
            ],
            "optional": false
        }
    ]
}
---

[TestInspectSuperComplexSchema - 2]
{
    "type": [
        "list"
    ],
    "of": [
        "string",
        "bottom"
    ],
    "properties": null
}
---

[TestInspectSuperComplexSchema - 3]
{
    "type": [
        "list",
        "struct"
    ],
    "of": null,
    "properties": []
}
---

[TestInspectSuperComplexSchema - 4]
{
    "type": [
        "list"
    ],
    "of": [
        "string",
        "bool",
        "int"
    ],
    "properties": null
}
---

[TestInspectSuperComplexSchema - 5]
{
    "type": [
        "list"
    ],
    "of": [
        "string",
        "bool",
        "int"
    ],
    "properties": null
}
---

[TestInspectSuperComplexSchema - 6]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "e",
                "a"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "e",
                "b"
            ],
            "type": [
                "string"
            ],
            "optional": false
        }
    ]
}
---

[TestInspectSuperComplexSchema - 7]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "e2",
                "a"
            ],
            "type": [
                "string"
            ],
            "optional": false
        },
        {
            "path": [
                "e2",
                "b"
            ],
            "type": [
                "string"
            ],
            "optional": false
        }
    ]
}
---

[TestInspectComplexSchema - 2]
{
    "type": [
        "bottom"
    ],
    "of": null,
    "properties": []
}
---

[TestInspectComplexSchema - 3]
{
    "type": [
        "bottom"
    ],
    "of": null,
    "properties": []
}
---

[TestInspectSuperComplexSchema - 8]
{
    "type": [
        "struct"
    ],
    "of": null,
    "properties": [
        {
            "path": [
                "f",
                "b"
            ],
            "type": [
                "string",
                "int"
            ],
            "optional": false
        }
    ]
}
---
